tket/tests/test_LexiRoute.cpp

```diff
@@ -454,10 +461,244 @@ SCENARIO("Test LexiRoute::solve") {
+       {nodes[1], nodes[2]},
+       {nodes[2], nodes[3]},
+       {nodes[3], nodes[4]}});
+  ArchitecturePtr shared_arc = std::make_shared<Architecture>(architecture);
+  GIVEN("No qubit to label, empty frontier, check_method.") {
+    Circuit circ(5);
+    std::shared_ptr<MappingFrontier> mf =
+        std::make_shared<MappingFrontier>(circ);
+    LabellingRoutingMethod lrm;
+    REQUIRE(!lrm.check_method(mf, shared_arc));
+  }
+  GIVEN("No qubit to label, partially filled frontier, check_method.") {
+    Circuit circ(5);
+    std::vector<Qubit> qubits = circ.all_qubits();
+    circ.add_op<UnitID>(OpType::CX, {qubits[0], qubits[4]});
+    circ.add_op<UnitID>(OpType::CZ, {qubits[1], qubits[2]});
+    circ.add_op<UnitID>(OpType::ZZPhase, 0.3, {qubits[3], qubits[0]});
+    std::map<UnitID, UnitID> rename_map = {
+        {qubits[0], nodes[0]},
+        {qubits[1], nodes[1]},
+        {qubits[2], nodes[2]},
+        {qubits[3], nodes[3]},
+        {qubits[4], nodes[4]}};
+    circ.rename_units(rename_map);
+    std::shared_ptr<MappingFrontier> mf =
+        std::make_shared<MappingFrontier>(circ);
+    LabellingRoutingMethod lrm;
+    REQUIRE(!lrm.check_method(mf, shared_arc));
+  }
+  GIVEN("Qubit to label, but casually restricted, check_method.") {
+    Circuit circ(5);
+    std::vector<Qubit> qubits = circ.all_qubits();
+    circ.add_op<UnitID>(OpType::CX, {qubits[0], qubits[4]});
+    circ.add_op<UnitID>(OpType::CZ, {qubits[1], qubits[2]});
+    circ.add_op<UnitID>(OpType::ZZPhase, 0.3, {qubits[3], qubits[0]});
+    std::map<UnitID, UnitID> rename_map = {
+        {qubits[0], nodes[0]},
+        {qubits[1], nodes[1]},
+        {qubits[2], nodes[2]},
+        {qubits[4], nodes[4]}};
+    circ.rename_units(rename_map);
+    std::shared_ptr<MappingFrontier> mf =
+        std::make_shared<MappingFrontier>(circ);
+    LabellingRoutingMethod lrm;
+    REQUIRE(!lrm.check_method(mf, shared_arc));
+  }
+  GIVEN(
+      "Two Qubit to label in future slice, causally restricted, "
+      "check_method.") {
+    Circuit circ(5);
+    std::vector<Qubit> qubits = circ.all_qubits();
+    circ.add_op<UnitID>(OpType::CX, {qubits[0], qubits[1]});
+    circ.add_op<UnitID>(OpType::CZ, {qubits[1], qubits[2]});
+    circ.add_op<UnitID>(OpType::CZ, {qubits[2], qubits[3]});
+    circ.add_op<UnitID>(OpType::ZZPhase, 0.3, {qubits[3], qubits[4]});
+    std::map<UnitID, UnitID> rename_map = {
+        {qubits[0], nodes[0]}, {qubits[1], nodes[1]}, {qubits[2], nodes[2]}};
+    circ.rename_units(rename_map);
+    std::shared_ptr<MappingFrontier> mf =
+        std::make_shared<MappingFrontier>(circ);
+    LabellingRoutingMethod lrm;
+    REQUIRE(!lrm.check_method(mf, shared_arc));
+  }
+  GIVEN("Three Qubit Gate, all labelled, first slice, check_method.") {
+    Circuit circ(5);
+    std::vector<Qubit> qubits = circ.all_qubits();
+    circ.add_op<UnitID>(OpType::CX, {qubits[0], qubits[4]});
+    circ.add_op<UnitID>(OpType::CCX, {qubits[1], qubits[2], qubits[3]});
```

Comment:
Is Lexiroute able to handle three qubit gates?

Reply:
Neither LexiRoute nor LexiLabelling can. I intend to add n-qubit gate labelling in the future (which was the original goal of this test), having deemed it too much work for this PR. However I kept this test case as it confirms that LexiLabelling check_method does not get thrown by n-qubit gates (it may be that we do n-qubit gate labelling with a different RoutingMethod so this would be an important trait).